blocks/basic/include/gnuradio-4.0/basic/SyncBlock.hpp

```diff
@@ -0,0 +1,363 @@
+)"">;
+
+template<typename T>
+struct SyncBlock : public gr::Block<SyncBlock<T>, NoDefaultTagForwarding, SyncBlockDoc> {
+    std::vector<gr::PortIn<T, gr::Async>> inputs;
+    std::vector<gr::PortOut<T>>           outputs;
+
+    // settings
+    Annotated<gr::Size_t, "n_ports", gr::Visible, gr::Doc<"variable number of in/out ports">, gr::Limits<1U, 32U>> n_ports          = 0U;
+    Annotated<gr::Size_t, "max_history_size", Doc<"Max size of history">>                                          max_history_size = 32000U; // should be less than actual buffer size (better < 80%)
+    Annotated<std::string, "filter", Doc<"trigger name filter">>                                                   filter           = "";
+    Annotated<std::uint64_t, "tolerance", Doc<"trigger time tolerance [ns]">>                                      tolerance        = 5ULL;
+
+    GR_MAKE_REFLECTABLE(SyncBlock, inputs, outputs, n_ports, max_history_size, filter, tolerance);
+
+    bool                     _isStreamSynchronized = false;
+    std::vector<std::size_t> _nDroppedSamples{}; // number of dropped samples, to be sent with desynchronized tag
+
+    int _processBulkCounter = 0;
+
+    struct SyncData {
+        std::size_t index; // sync index
+        std::size_t nPre;  // number of pre samples, sample with sync index is not included
+        std::size_t nPost; // number of post samples, sample with sync index is not included
+    };
+
+    void settingsChanged(const property_map& oldSettings, const property_map& newSettings) {
+        if (newSettings.contains("n_ports") && oldSettings.at("n_ports") != newSettings.at("n_ports")) {
+            // if one of the port is already connected and n_ports was changed then throw
+            bool inConnected  = std::any_of(inputs.begin(), inputs.end(), [](const auto& port) { return port.isConnected(); });
+            bool outConnected = std::any_of(outputs.begin(), outputs.end(), [](const auto& port) { return port.isConnected(); });
+            if (inConnected || outConnected) {
+                using namespace gr::message;
+                std::string messageError{"Number of input/output ports cannot be changed after Graph initialization."};
+                // this->emitMessage(this->msgOut, { { key::Kind, kind::Error }, { key::ErrorInfo, messageError } });
+                throw std::range_error(messageError);
+            }
+            fmt::print("{}: configuration changed: n_ports {} -> {}\n", this->name, oldSettings.at("n_ports"), newSettings.at("n_ports"));
+            inputs.resize(n_ports);
+            outputs.resize(n_ports);
+            _nDroppedSamples.resize(n_ports, 0UZ);
+        }
+
+        if (newSettings.contains("max_history_size")) {
+            // should be less than actual buffer size (better < 80%)
+            // The logic has to be implemented when the new Port API is available
+        }
```

Comment:
As discussed in person: the block implementation should not assume nor use the Port's buffer API. 

I think it's OK to define a `max_history_size` that is used to define/communicate the required minimum buffer size to the graph and scheduler, and in turn be used implicitly by the block as the buffer having at least `max_history_size` slots.

However, the block should not depend on the size of the actually used Port-buffer. To note, we want to keep the buffer API slim and also connect blocks using other mechanisms than via buffers (<-> merge API).

Reply:
The idea here is to check if max_history_size is actually < buffer size, if not throw an gr::exception.